src/spawn_canister/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use ic_cdk::export::candid::{CandidType, Deserialize, Principal, Encode};
+use ic_cdk_macros::{update};
+use ic_cdk::api::call;
+
+#[derive(CandidType, Deserialize)]
+pub enum CreateCanisterResponse{
+    #[serde(rename = "created")]
+    Created(Principal),
+
+    #[serde(rename = "insufficient_funds")]
+    InsufficientFunds,
+
+    #[serde(rename = "canister_creation_error")]
+    CanisterCreationError,
+
+    #[serde(rename = "canister_installation_error")]
+    CanisterInstallationError,
+
+    #[serde(rename = "change_controller_error")]
+    ChangeControllerError,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct CreateCanisterResult {
+    pub canister_id: Principal,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum InstallMode {
+    #[serde(rename = "install")]
+    Install,
+    #[serde(rename = "reinstall")]
+    Reinstall,
+    #[serde(rename = "upgrade")]
+    Upgrade,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct InstallCodeArg {
+    pub mode: InstallMode,
+    pub canister_id: Principal,
+    pub wasm_module: Vec<u8>,
+    pub arg : Vec<u8>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct CanisterSettings{
+    controllers: Option<Vec<Principal>>,
+    compute_allocation: Option<candid::Nat>,
+    memory_allocation: Option<candid::Nat>,
+    freezing_threshold: Option<candid::Nat>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct UpdateSettingsArg{
+    canister_id: Principal,
+    settings: CanisterSettings,
+}
+
+static VIDEO_CANISTER_CODE: &[u8;  include_bytes!("../../../target/wasm32-unknown-unknown/release/video_canister_opt.wasm").len()] = include_bytes!("../../../target/wasm32-unknown-unknown/release/video_canister_opt.wasm");
+
+const MIN_CANISTER_CYCLES_REQUIRED: u64 = 100_000_000_000 + 10_000_000_000; //TODO rough guess, calculate correct costs
+
+#[update]
+pub async fn create_new_canister() -> CreateCanisterResponse{
+
+    let owner = ic_cdk::api::caller(); //call early before any callbacks from calling other canisters
+    let available_cycles = call::msg_cycles_available();  
+
+    if available_cycles < MIN_CANISTER_CYCLES_REQUIRED{
+        return CreateCanisterResponse::InsufficientFunds;
+    } else {
+        call::msg_cycles_accept(available_cycles);
+    }
+
+    let canister_princ = match create_canister_on_network().await{
+        Ok(new_princ) => new_princ,
```

Comment:
Koennte man hier nicht wie unten (Zeile 83 z.b.) "if let Err(..." schreiben? mit new_princ passiert ja sowieso nix

Reply:
Nein, new_princ wird ausgepackt, zurückgegeben und in canister_princ geschrieben. Ist das zu unklar? Andere Lösungen fände ich unschöner

Reply:
Passt, ist fehlendes Rust wissen von mir!